src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -160,16 +166,214 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
             if (composite == null)
             {
-                builder.Add(text);
+                segments.Add(text);
             }
             else
             {
-                builder.AddRange(composite._texts);
+                segments.AddRange(composite._segments);
+            }
+        }
+
+        internal static SourceText ToSourceTextAndFree(ArrayBuilder<SourceText> segments, SourceText original, bool adjustSegments)
+        {
+            if (adjustSegments)
+            {
+                TrimInaccessibleText(segments);
+                ReduceSegmentCount(segments);
+            }
+
+            if (segments.Count == 0)
+            {
+                segments.Free();
+                return SourceText.From(string.Empty, original.Encoding, original.ChecksumAlgorithm);
+            }
+            else if (segments.Count == 1)
+            {
+                SourceText result = segments[0];
+                segments.Free();
+                return result;
+            }
+            else
+            {
+                return new CompositeText(segments.ToImmutableAndFree(), original.Encoding, original.ChecksumAlgorithm);
+            }
+        }
+
+        internal const int IDEAL_SEGMENT_COUNT = 10;
+        internal const int MAXIMUM_SEGMENT_COUNT = 20;
+
+        /// <summary>
+        /// Reduces the number of segments toward the ideal number of segments.
+        /// </summary>
+        private static void ReduceSegmentCount(ArrayBuilder<SourceText> segments)
+        {
+            if (segments.Count > MAXIMUM_SEGMENT_COUNT)
+            {
+                var minimumSegmentSize = GetMinimalSegmentSize(segments);
+                if (minimumSegmentSize > 0)
+                {
+                    CombineSegments(segments, minimumSegmentSize);
+                }
+            }
+        }
+
+        private const int INITIAL_SEGMENT_SIZE = 32;        // allow combining if each text segment size is less than this
+        private const int MAXIMUM_SEGMENT_SIZE = int.MaxValue / 16; // give up on reducing if segment sizes are all greater than this
+
+        /// <summary>
+        /// Gets the minimal segment size used to determine which segments are eligible to be combined with their neighbors.
+        /// </summary>
+        private static int GetMinimalSegmentSize(ArrayBuilder<SourceText> segments)
+        {
+            // find the minimal segment size that reduces enough segments to less that or equal to the ideal segment count
+            for (var minimumSegmentSize = INITIAL_SEGMENT_SIZE;
+                minimumSegmentSize <= MAXIMUM_SEGMENT_SIZE;
+                minimumSegmentSize *= 2)
+            {
+                if (GetReducedSize(segments, minimumSegmentSize) <= IDEAL_SEGMENT_COUNT)
+                {
+                    return minimumSegmentSize;
+                }
+            }
+
+            // cannot find an appropriate size for segments to be combined.
+            return -1;
```

Comment:
The > 0 check in the caller above implied it could be. So perhaps that simplifies?